{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d133e701",
   "metadata": {},
   "source": [
    "# Gunc simulated genomes type 3b: determine appropriate taxonomic level for each contiguous sequence in synthetic genomes\n",
    "\n",
    "The gunc genomes were simulated from genomes, using the NCBI taxonomy to determine the taxonomic level at which the contamination deviated from the dominate lineage. \n",
    "Since we're using GTDB taxonomy, this notebook determines the taxonomic level at which the contaminant genomes different from the dominant genome within the GTDB taxonomy.\n",
    "\n",
    "The output is saved as a tsv file to `sandbox/gunc_synthetic_genomes3b/type3b.genomes/genome_contamination_level.tsv`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "996daa6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "setwd(\"..\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed192443",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(readr)\n",
    "library(tidyr)\n",
    "library(purrr)\n",
    "library(rentrez)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5780855f",
   "metadata": {},
   "source": [
    "## obtain and read in GTDB rs207 metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8aa262da",
   "metadata": {},
   "outputs": [],
   "source": [
    "destfile <- \"inputs/gtdb-rs207/bac120_metadata_rs207.tar.gz\"\n",
    "url <- \"https://data.gtdb.ecogenomic.org/releases/release207/207.0/bac120_metadata_r207.tar.gz\"\n",
    "if (!file.exists(destfile)) {\n",
    "  download.file(url, destfile, method=\"auto\") \n",
    "}\n",
    "outfile <- \"inputs/gtdb-rs207/bac120_metadata_r207.tsv\"\n",
    "if (!file.exists(outfile)){\n",
    "  untar(destfile, exdir = \"inputs/gtdb-rs207\")\n",
    "}\n",
    "\n",
    "destfile <- \"inputs/gtdb-rs207/ar53_metadata_rs207.tar.gz\"\n",
    "url <- \"https://data.gtdb.ecogenomic.org/releases/release207/207.0/ar53_metadata_r207.tar.gz\"\n",
    "if (!file.exists(destfile)) {\n",
    "  download.file(url, destfile, method=\"auto\") \n",
    "}\n",
    "outfile <- \"inputs/gtdb-r207/ar53_metadata_r207.tsv\"\n",
    "if (!file.exists(outfile)){\n",
    "  untar(destfile, exdir = \"inputs/gtdb-rs207\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2ad84ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“One or more parsing issues, see `problems()` for details”\n"
     ]
    }
   ],
   "source": [
    "gtdb_metadata_rs207 <- read_tsv(\"inputs/gtdb-rs207/bac120_metadata_r207.tsv\", show_col_types = FALSE) %>%\n",
    "  #separate(gtdb_taxonomy, into = c(\"domain\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"), \n",
    "  #         sep = \";\") %>%\n",
    "  mutate(ncbi_database = ifelse(grepl(pattern = \"GB\", accession), \"GenBank\", \"none\"),\n",
    "         ncbi_database = ifelse(grepl(pattern = \"RS\", accession), \"RefSeq\", ncbi_database)) %>%\n",
    "  mutate(accession = gsub(\"^RS_\", \"\", accession),\n",
    "         accession = gsub(\"^GB_\", \"\", accession),\n",
    "         accession_pruned = gsub(\"GC[FA]_\", \"\", accession),\n",
    "         accession_pruned = gsub(\"\\\\..\", \"\", accession_pruned))\n",
    "\n",
    "gtdb_metadata_rs207 <- read_tsv(\"inputs/gtdb-rs207/ar53_metadata_r207.tsv\", show_col_types = FALSE) %>%\n",
    "  #separate(gtdb_taxonomy, into = c(\"domain\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"), \n",
    "  #         sep = \";\") %>%\n",
    "  mutate(ncbi_database = ifelse(grepl(pattern = \"GB\", accession), \"GenBank\", \"none\"),\n",
    "         ncbi_database = ifelse(grepl(pattern = \"RS\", accession), \"RefSeq\", ncbi_database)) %>%\n",
    "  mutate(accession = gsub(\"^RS_\", \"\", accession),\n",
    "         accession = gsub(\"^GB_\", \"\", accession),\n",
    "         accession_pruned = gsub(\"GC[FA]_\", \"\", accession),\n",
    "         accession_pruned = gsub(\"\\\\..\", \"\", accession_pruned)) %>%\n",
    "  bind_rows(gtdb_metadata_rs207)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee55762a",
   "metadata": {},
   "source": [
    "## read in and parse faidx files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3e33b586",
   "metadata": {},
   "outputs": [],
   "source": [
    "# faidx colname legend\n",
    "# NAME\tName of this reference sequence\n",
    "# LENGTH\tTotal length of this reference sequence, in bases\n",
    "# OFFSET\tOffset in the FASTA/FASTQ file of this sequence's first base\n",
    "# LINEBASES\tThe number of bases on each line\n",
    "# LINEWIDTH\tThe number of bytes in each line, including the newline\n",
    "# QUALOFFSET\tOffset of sequence's first quality within the FASTQ file\n",
    "\n",
    "faidx_rs207_reps <- Sys.glob(\"sandbox/gunc_synthetic_genomes3b/type3b.genomes/outputs/genomes_to_charcoal3_vs_gtdb_rs207_reps/*fai\") %>%\n",
    "  set_names() %>%\n",
    "  map_dfr(read_tsv, show_col_types = F, col_names = c(\"contig_name\", \"length\", \"offset\", \"linebases\", \"linewidth\", \"qualoffset\"), .id = \"filename\") %>%\n",
    "  mutate(type = ifelse(grepl(\"clean\", x = filename), \"clean\", \"dirty\"),\n",
    "         genome = gsub(\"_\\\\.fa.*\", \"\", basename(filename))) %>%\n",
    "  select(genome, type, contig_name, length) %>%\n",
    "  separate(contig_name, into = c(\"num\", \"genome_accession\", \"contig\"), sep = \"\\\\.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd12de39",
   "metadata": {},
   "source": [
    "## convert biosample ID to genbank/refseq accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "304469b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get conversion between biosample accession and assembly accession\n",
    "biosample_to_assembly_df <- data.frame()\n",
    "for(biosample_accession in unique(faidx_rs207_reps$genome_accession)){\n",
    "    tmp_search <- entrez_search(db=\"assembly\", term= biosample_accession) # search for the id assoc with this record\n",
    "    tmp_summary <- entrez_summary(db=\"assembly\", id=tmp_search$id)        # use id to retrieve info about accession\n",
    "    if(length(tmp_summary) > 5){ # assemblies with multiple accessions will have nested lists, with the first list being shallow\n",
    "        assembly_accession <- tmp_summary$assemblyaccession               # extract assembly accession\n",
    "        tmp_biosample_to_assembly_df <- data.frame(biosample = biosample_accession, assembly = assembly_accession) # make temp df \n",
    "        biosample_to_assembly_df <- bind_rows(biosample_to_assembly_df, tmp_biosample_to_assembly_df) # bind to final results\n",
    "    } else {\n",
    "        assembly_accession <- tmp_summary[[1]]$assemblyaccession          # extract assembly accession\n",
    "        tmp_biosample_to_assembly_df <- data.frame(biosample = biosample_accession, assembly = assembly_accession) # make temp df \n",
    "        biosample_to_assembly_df <- bind_rows(biosample_to_assembly_df, tmp_biosample_to_assembly_df) # bind to final results\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "23b97a1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# join to gtdb lineage information\n",
    "faidx_rs207_reps_annotated <- faidx_rs207_reps %>%\n",
    "  left_join(biosample_to_assembly_df, by = c(\"genome_accession\" = \"biosample\")) %>%\n",
    "  mutate(assembly_pruned =  gsub(\"GC[AF]_\", \"\", assembly),\n",
    "         assembly_pruned = gsub(\"\\\\..\", \"\", assembly_pruned)) %>%\n",
    "  left_join(gtdb_metadata_rs207, by = c(\"assembly_pruned\" = \"accession_pruned\")) %>%\n",
    "  mutate(contamination_level = gsub(\"type3b_\", \"\", genome),\n",
    "         contamination_level = gsub(\"_.*\", \"\", contamination_level)) %>%\n",
    "  rename(contig_lineage = gtdb_taxonomy)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "64216fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# join to gather lineage info and label as correct lineage, contaminate genome1, or contaminate genome2\n",
    "\n",
    "lineages_rs207 <- read_csv(\"sandbox/gunc_synthetic_genomes3b/type3b.genomes/genomes_to_charcoal3_rs207_lineages.csv\", show_col_types = F) %>%\n",
    "  mutate(dom_lineage = paste(superkingdom, phylum, class, order, family, genus, species, sep = \";\")) %>%\n",
    "  mutate(ident = gsub(\"_\\\\.fa\", \"\", ident)) %>%\n",
    "  select(ident, dom_lineage)\n",
    "\n",
    "faidx_rs207_reps_annotated <- faidx_rs207_reps_annotated %>%\n",
    "  left_join(lineages_rs207, by = c(\"genome\" = \"ident\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd752950",
   "metadata": {},
   "source": [
    "## Re-infer what levels of taxonomy contamination varies based on GTDB \n",
    "\n",
    "I think levels were previously set based on NCBI taxonomy. Convert to GTDB taxonomy to be copacetic with other analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "482cdbca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# figure out which taxonomic level the dom lineage is first mismatched from the contam lineage\n",
    "reset_contam_level1 <- faidx_rs207_reps_annotated %>%\n",
    "  select(genome, contamination_level, dom_lineage, contig_lineage) %>%\n",
    "  distinct() %>%\n",
    "  separate(dom_lineage, into = c(\"dom_domain\", \"dom_phylum\", \"dom_class\", \"dom_order\", \"dom_family\", \"dom_genus\", \"dom_species\"), sep = \";\") %>%\n",
    "  separate(contig_lineage, into = c(\"contig_domain\", \"contig_phylum\", \"contig_class\", \"contig_order\", \"contig_family\", \"contig_genus\", \"contig_species\"), sep = \";\") %>%\n",
    "  group_by(genome) %>%\n",
    "  mutate(gtdb_contamination_level = ifelse(dom_phylum != contig_phylum, \"phylum\", \n",
    "                                           ifelse(dom_class != contig_class, \"class\", \n",
    "                                                  ifelse(dom_order != contig_order, \"order\",\n",
    "                                                         ifelse(dom_family != contig_family, \"family\",\n",
    "                                                               ifelse(dom_genus != contig_genus, \"genus\",\n",
    "                                                                     ifelse(dom_species != contig_species, \"species\", \"dom_lineage\")))))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "51394b0d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 60 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>genome</th><th scope=col>contamination_level</th><th scope=col>gtdb_contamination_level</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1121877.SAMN02745886 </td><td>class </td><td>class  </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1122947.SAMN01057329 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1123249.SAMN02440643 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1149862.SAMN01057330 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1160201.SAMN00811184 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1231054.SAMN01730125 </td><td>class </td><td>class  </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1231376.SAMN02469699 </td><td>class </td><td>class  </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1287313.SAMN02359689 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1298595.SAMD00000528 </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1505.SAMEA1572114    </td><td>class </td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1051653.SAMN02597184</td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1278077.SAMN02469985</td><td>family</td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1339247.SAMN02767912</td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1352.SAMEA3893589   </td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1423761.SAMN02369403</td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1423802.SAMD00000539</td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1428.SAMN04278951   </td><td>family</td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1428425.SAMN02904288</td><td>family</td><td>family </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1433.SAMN05934573   </td><td>family</td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1502.SAMN05323905   </td><td>family</td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_112002.SAMN05660405  </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1120962.SAMN02441201 </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1226750.SAMN04537322 </td><td>genus </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1298610.SAMD00000012 </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1305781.PRJEB1633    </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1335621.SAMN04537321 </td><td>genus </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1370125.SAMN02471859 </td><td>genus </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1391646.SAMN02324641 </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1409964.SAMN03486520 </td><td>genus </td><td>genus  </td></tr>\n",
       "\t<tr><td>type3b_genus_3_0.33_0000_1421.SAMN04215005    </td><td>genus </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1050199.SAMN02261318 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1136162.SAMN02470508 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_115545.SAMN06140942  </td><td>order </td><td>family </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1158154.SAMN02441053 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_119857.SAMN03773898  </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1311022.SAMN02644703 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1566202.SAMN03159405 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1648182.SAMN03571527 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1735692.SAMN04151589 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_1798220.SAMN04515685 </td><td>order </td><td>order  </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1001707.SAMN00672618</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1188240.SAMN02471343</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1280.SAMEA862559    </td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1280.SAMN04018804   </td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1313.SAMEA1025997   </td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1316911.SAMN02726251</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1354152.SAMN02231122</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1410673.SAMN02841180</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1413333.SAMN02391160</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_phylum_3_0.33_0000_1447442.SAMN02584593</td><td>phylum</td><td>phylum </td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1088883.SAMN00993432 </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1144667.SAMN01828172 </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1276282.SAMN02603493 </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1283.SAMEA1035032    </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1283.SAMN04361561    </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1307.SAMEA1566015    </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1307.SAMEA1566063    </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1310908.SAMN02570339 </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1311.SAMEA2780151    </td><td>specI </td><td>species</td></tr>\n",
       "\t<tr><td>type3b_specI_3_0.33_0000_1311.SAMN05762126    </td><td>specI </td><td>species</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 60 × 3\n",
       "\\begin{tabular}{lll}\n",
       " genome & contamination\\_level & gtdb\\_contamination\\_level\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1121877.SAMN02745886  & class  & class  \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1122947.SAMN01057329  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1123249.SAMN02440643  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1149862.SAMN01057330  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1160201.SAMN00811184  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1231054.SAMN01730125  & class  & class  \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1231376.SAMN02469699  & class  & class  \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1287313.SAMN02359689  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1298595.SAMD00000528  & class  & phylum \\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1505.SAMEA1572114     & class  & phylum \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1051653.SAMN02597184 & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1278077.SAMN02469985 & family & genus  \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1339247.SAMN02767912 & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1352.SAMEA3893589    & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1423761.SAMN02369403 & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1423802.SAMD00000539 & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1428.SAMN04278951    & family & order  \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1428425.SAMN02904288 & family & family \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1433.SAMN05934573    & family & order  \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1502.SAMN05323905    & family & order  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_112002.SAMN05660405   & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1120962.SAMN02441201  & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1226750.SAMN04537322  & genus  & species\\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1298610.SAMD00000012  & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1305781.PRJEB1633     & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1335621.SAMN04537321  & genus  & species\\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1370125.SAMN02471859  & genus  & species\\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1391646.SAMN02324641  & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1409964.SAMN03486520  & genus  & genus  \\\\\n",
       "\t type3b\\_genus\\_3\\_0.33\\_0000\\_1421.SAMN04215005     & genus  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1050199.SAMN02261318  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1136162.SAMN02470508  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_115545.SAMN06140942   & order  & family \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1158154.SAMN02441053  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_119857.SAMN03773898   & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1311022.SAMN02644703  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1566202.SAMN03159405  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1648182.SAMN03571527  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1735692.SAMN04151589  & order  & order  \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_1798220.SAMN04515685  & order  & order  \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1001707.SAMN00672618 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1188240.SAMN02471343 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1280.SAMEA862559     & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1280.SAMN04018804    & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1313.SAMEA1025997    & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1316911.SAMN02726251 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1354152.SAMN02231122 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1410673.SAMN02841180 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1413333.SAMN02391160 & phylum & phylum \\\\\n",
       "\t type3b\\_phylum\\_3\\_0.33\\_0000\\_1447442.SAMN02584593 & phylum & phylum \\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1088883.SAMN00993432  & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1144667.SAMN01828172  & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1276282.SAMN02603493  & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1283.SAMEA1035032     & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1283.SAMN04361561     & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1307.SAMEA1566015     & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1307.SAMEA1566063     & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1310908.SAMN02570339  & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1311.SAMEA2780151     & specI  & species\\\\\n",
       "\t type3b\\_specI\\_3\\_0.33\\_0000\\_1311.SAMN05762126     & specI  & species\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 60 × 3\n",
       "\n",
       "| genome &lt;chr&gt; | contamination_level &lt;chr&gt; | gtdb_contamination_level &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| type3b_class_3_0.33_0000_1121877.SAMN02745886  | class  | class   |\n",
       "| type3b_class_3_0.33_0000_1122947.SAMN01057329  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1123249.SAMN02440643  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1149862.SAMN01057330  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1160201.SAMN00811184  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1231054.SAMN01730125  | class  | class   |\n",
       "| type3b_class_3_0.33_0000_1231376.SAMN02469699  | class  | class   |\n",
       "| type3b_class_3_0.33_0000_1287313.SAMN02359689  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1298595.SAMD00000528  | class  | phylum  |\n",
       "| type3b_class_3_0.33_0000_1505.SAMEA1572114     | class  | phylum  |\n",
       "| type3b_family_3_0.33_0000_1051653.SAMN02597184 | family | family  |\n",
       "| type3b_family_3_0.33_0000_1278077.SAMN02469985 | family | genus   |\n",
       "| type3b_family_3_0.33_0000_1339247.SAMN02767912 | family | family  |\n",
       "| type3b_family_3_0.33_0000_1352.SAMEA3893589    | family | family  |\n",
       "| type3b_family_3_0.33_0000_1423761.SAMN02369403 | family | family  |\n",
       "| type3b_family_3_0.33_0000_1423802.SAMD00000539 | family | family  |\n",
       "| type3b_family_3_0.33_0000_1428.SAMN04278951    | family | order   |\n",
       "| type3b_family_3_0.33_0000_1428425.SAMN02904288 | family | family  |\n",
       "| type3b_family_3_0.33_0000_1433.SAMN05934573    | family | order   |\n",
       "| type3b_family_3_0.33_0000_1502.SAMN05323905    | family | order   |\n",
       "| type3b_genus_3_0.33_0000_112002.SAMN05660405   | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1120962.SAMN02441201  | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1226750.SAMN04537322  | genus  | species |\n",
       "| type3b_genus_3_0.33_0000_1298610.SAMD00000012  | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1305781.PRJEB1633     | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1335621.SAMN04537321  | genus  | species |\n",
       "| type3b_genus_3_0.33_0000_1370125.SAMN02471859  | genus  | species |\n",
       "| type3b_genus_3_0.33_0000_1391646.SAMN02324641  | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1409964.SAMN03486520  | genus  | genus   |\n",
       "| type3b_genus_3_0.33_0000_1421.SAMN04215005     | genus  | order   |\n",
       "| type3b_order_3_0.33_0000_1050199.SAMN02261318  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1136162.SAMN02470508  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_115545.SAMN06140942   | order  | family  |\n",
       "| type3b_order_3_0.33_0000_1158154.SAMN02441053  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_119857.SAMN03773898   | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1311022.SAMN02644703  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1566202.SAMN03159405  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1648182.SAMN03571527  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1735692.SAMN04151589  | order  | order   |\n",
       "| type3b_order_3_0.33_0000_1798220.SAMN04515685  | order  | order   |\n",
       "| type3b_phylum_3_0.33_0000_1001707.SAMN00672618 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1188240.SAMN02471343 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1280.SAMEA862559     | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1280.SAMN04018804    | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1313.SAMEA1025997    | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1316911.SAMN02726251 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1354152.SAMN02231122 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1410673.SAMN02841180 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1413333.SAMN02391160 | phylum | phylum  |\n",
       "| type3b_phylum_3_0.33_0000_1447442.SAMN02584593 | phylum | phylum  |\n",
       "| type3b_specI_3_0.33_0000_1088883.SAMN00993432  | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1144667.SAMN01828172  | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1276282.SAMN02603493  | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1283.SAMEA1035032     | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1283.SAMN04361561     | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1307.SAMEA1566015     | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1307.SAMEA1566063     | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1310908.SAMN02570339  | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1311.SAMEA2780151     | specI  | species |\n",
       "| type3b_specI_3_0.33_0000_1311.SAMN05762126     | specI  | species |\n",
       "\n"
      ],
      "text/plain": [
       "   genome                                         contamination_level\n",
       "1  type3b_class_3_0.33_0000_1121877.SAMN02745886  class              \n",
       "2  type3b_class_3_0.33_0000_1122947.SAMN01057329  class              \n",
       "3  type3b_class_3_0.33_0000_1123249.SAMN02440643  class              \n",
       "4  type3b_class_3_0.33_0000_1149862.SAMN01057330  class              \n",
       "5  type3b_class_3_0.33_0000_1160201.SAMN00811184  class              \n",
       "6  type3b_class_3_0.33_0000_1231054.SAMN01730125  class              \n",
       "7  type3b_class_3_0.33_0000_1231376.SAMN02469699  class              \n",
       "8  type3b_class_3_0.33_0000_1287313.SAMN02359689  class              \n",
       "9  type3b_class_3_0.33_0000_1298595.SAMD00000528  class              \n",
       "10 type3b_class_3_0.33_0000_1505.SAMEA1572114     class              \n",
       "11 type3b_family_3_0.33_0000_1051653.SAMN02597184 family             \n",
       "12 type3b_family_3_0.33_0000_1278077.SAMN02469985 family             \n",
       "13 type3b_family_3_0.33_0000_1339247.SAMN02767912 family             \n",
       "14 type3b_family_3_0.33_0000_1352.SAMEA3893589    family             \n",
       "15 type3b_family_3_0.33_0000_1423761.SAMN02369403 family             \n",
       "16 type3b_family_3_0.33_0000_1423802.SAMD00000539 family             \n",
       "17 type3b_family_3_0.33_0000_1428.SAMN04278951    family             \n",
       "18 type3b_family_3_0.33_0000_1428425.SAMN02904288 family             \n",
       "19 type3b_family_3_0.33_0000_1433.SAMN05934573    family             \n",
       "20 type3b_family_3_0.33_0000_1502.SAMN05323905    family             \n",
       "21 type3b_genus_3_0.33_0000_112002.SAMN05660405   genus              \n",
       "22 type3b_genus_3_0.33_0000_1120962.SAMN02441201  genus              \n",
       "23 type3b_genus_3_0.33_0000_1226750.SAMN04537322  genus              \n",
       "24 type3b_genus_3_0.33_0000_1298610.SAMD00000012  genus              \n",
       "25 type3b_genus_3_0.33_0000_1305781.PRJEB1633     genus              \n",
       "26 type3b_genus_3_0.33_0000_1335621.SAMN04537321  genus              \n",
       "27 type3b_genus_3_0.33_0000_1370125.SAMN02471859  genus              \n",
       "28 type3b_genus_3_0.33_0000_1391646.SAMN02324641  genus              \n",
       "29 type3b_genus_3_0.33_0000_1409964.SAMN03486520  genus              \n",
       "30 type3b_genus_3_0.33_0000_1421.SAMN04215005     genus              \n",
       "31 type3b_order_3_0.33_0000_1050199.SAMN02261318  order              \n",
       "32 type3b_order_3_0.33_0000_1136162.SAMN02470508  order              \n",
       "33 type3b_order_3_0.33_0000_115545.SAMN06140942   order              \n",
       "34 type3b_order_3_0.33_0000_1158154.SAMN02441053  order              \n",
       "35 type3b_order_3_0.33_0000_119857.SAMN03773898   order              \n",
       "36 type3b_order_3_0.33_0000_1311022.SAMN02644703  order              \n",
       "37 type3b_order_3_0.33_0000_1566202.SAMN03159405  order              \n",
       "38 type3b_order_3_0.33_0000_1648182.SAMN03571527  order              \n",
       "39 type3b_order_3_0.33_0000_1735692.SAMN04151589  order              \n",
       "40 type3b_order_3_0.33_0000_1798220.SAMN04515685  order              \n",
       "41 type3b_phylum_3_0.33_0000_1001707.SAMN00672618 phylum             \n",
       "42 type3b_phylum_3_0.33_0000_1188240.SAMN02471343 phylum             \n",
       "43 type3b_phylum_3_0.33_0000_1280.SAMEA862559     phylum             \n",
       "44 type3b_phylum_3_0.33_0000_1280.SAMN04018804    phylum             \n",
       "45 type3b_phylum_3_0.33_0000_1313.SAMEA1025997    phylum             \n",
       "46 type3b_phylum_3_0.33_0000_1316911.SAMN02726251 phylum             \n",
       "47 type3b_phylum_3_0.33_0000_1354152.SAMN02231122 phylum             \n",
       "48 type3b_phylum_3_0.33_0000_1410673.SAMN02841180 phylum             \n",
       "49 type3b_phylum_3_0.33_0000_1413333.SAMN02391160 phylum             \n",
       "50 type3b_phylum_3_0.33_0000_1447442.SAMN02584593 phylum             \n",
       "51 type3b_specI_3_0.33_0000_1088883.SAMN00993432  specI              \n",
       "52 type3b_specI_3_0.33_0000_1144667.SAMN01828172  specI              \n",
       "53 type3b_specI_3_0.33_0000_1276282.SAMN02603493  specI              \n",
       "54 type3b_specI_3_0.33_0000_1283.SAMEA1035032     specI              \n",
       "55 type3b_specI_3_0.33_0000_1283.SAMN04361561     specI              \n",
       "56 type3b_specI_3_0.33_0000_1307.SAMEA1566015     specI              \n",
       "57 type3b_specI_3_0.33_0000_1307.SAMEA1566063     specI              \n",
       "58 type3b_specI_3_0.33_0000_1310908.SAMN02570339  specI              \n",
       "59 type3b_specI_3_0.33_0000_1311.SAMEA2780151     specI              \n",
       "60 type3b_specI_3_0.33_0000_1311.SAMN05762126     specI              \n",
       "   gtdb_contamination_level\n",
       "1  class                   \n",
       "2  phylum                  \n",
       "3  phylum                  \n",
       "4  phylum                  \n",
       "5  phylum                  \n",
       "6  class                   \n",
       "7  class                   \n",
       "8  phylum                  \n",
       "9  phylum                  \n",
       "10 phylum                  \n",
       "11 family                  \n",
       "12 genus                   \n",
       "13 family                  \n",
       "14 family                  \n",
       "15 family                  \n",
       "16 family                  \n",
       "17 order                   \n",
       "18 family                  \n",
       "19 order                   \n",
       "20 order                   \n",
       "21 genus                   \n",
       "22 genus                   \n",
       "23 species                 \n",
       "24 genus                   \n",
       "25 genus                   \n",
       "26 species                 \n",
       "27 species                 \n",
       "28 genus                   \n",
       "29 genus                   \n",
       "30 order                   \n",
       "31 order                   \n",
       "32 order                   \n",
       "33 family                  \n",
       "34 order                   \n",
       "35 order                   \n",
       "36 order                   \n",
       "37 order                   \n",
       "38 order                   \n",
       "39 order                   \n",
       "40 order                   \n",
       "41 phylum                  \n",
       "42 phylum                  \n",
       "43 phylum                  \n",
       "44 phylum                  \n",
       "45 phylum                  \n",
       "46 phylum                  \n",
       "47 phylum                  \n",
       "48 phylum                  \n",
       "49 phylum                  \n",
       "50 phylum                  \n",
       "51 species                 \n",
       "52 species                 \n",
       "53 species                 \n",
       "54 species                 \n",
       "55 species                 \n",
       "56 species                 \n",
       "57 species                 \n",
       "58 species                 \n",
       "59 species                 \n",
       "60 species                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "reset_contam_level2 <- reset_contam_level1 %>%\n",
    "  filter(gtdb_contamination_level != \"dom_lineage\") %>%             # remove where dom == dom\n",
    "  select(genome, contamination_level, gtdb_contamination_level) %>% # select genome and contam levels\n",
    "  group_by(genome) %>%                                              # group by genome\n",
    "  arrange(genome, factor(gtdb_contamination_level, levels = c(\"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"))) %>% # arrange by lineage order\n",
    "  slice_head(n = 1)                                                 # select top lineage difference\n",
    "\n",
    "  \n",
    "reset_contam_level2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4918b34a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A grouped_df: 12 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>genome</th><th scope=col>contamination_level</th><th scope=col>gtdb_contamination_level</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1122947.SAMN01057329 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1123249.SAMN02440643 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1149862.SAMN01057330 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1160201.SAMN00811184 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1287313.SAMN02359689 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1298595.SAMD00000528 </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_class_3_0.33_0000_1505.SAMEA1572114    </td><td>class </td><td>phylum</td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1278077.SAMN02469985</td><td>family</td><td>genus </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1428.SAMN04278951   </td><td>family</td><td>order </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1433.SAMN05934573   </td><td>family</td><td>order </td></tr>\n",
       "\t<tr><td>type3b_family_3_0.33_0000_1502.SAMN05323905   </td><td>family</td><td>order </td></tr>\n",
       "\t<tr><td>type3b_order_3_0.33_0000_115545.SAMN06140942  </td><td>order </td><td>family</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A grouped\\_df: 12 × 3\n",
       "\\begin{tabular}{lll}\n",
       " genome & contamination\\_level & gtdb\\_contamination\\_level\\\\\n",
       " <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1122947.SAMN01057329  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1123249.SAMN02440643  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1149862.SAMN01057330  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1160201.SAMN00811184  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1287313.SAMN02359689  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1298595.SAMD00000528  & class  & phylum\\\\\n",
       "\t type3b\\_class\\_3\\_0.33\\_0000\\_1505.SAMEA1572114     & class  & phylum\\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1278077.SAMN02469985 & family & genus \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1428.SAMN04278951    & family & order \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1433.SAMN05934573    & family & order \\\\\n",
       "\t type3b\\_family\\_3\\_0.33\\_0000\\_1502.SAMN05323905    & family & order \\\\\n",
       "\t type3b\\_order\\_3\\_0.33\\_0000\\_115545.SAMN06140942   & order  & family\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A grouped_df: 12 × 3\n",
       "\n",
       "| genome &lt;chr&gt; | contamination_level &lt;chr&gt; | gtdb_contamination_level &lt;chr&gt; |\n",
       "|---|---|---|\n",
       "| type3b_class_3_0.33_0000_1122947.SAMN01057329  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1123249.SAMN02440643  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1149862.SAMN01057330  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1160201.SAMN00811184  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1287313.SAMN02359689  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1298595.SAMD00000528  | class  | phylum |\n",
       "| type3b_class_3_0.33_0000_1505.SAMEA1572114     | class  | phylum |\n",
       "| type3b_family_3_0.33_0000_1278077.SAMN02469985 | family | genus  |\n",
       "| type3b_family_3_0.33_0000_1428.SAMN04278951    | family | order  |\n",
       "| type3b_family_3_0.33_0000_1433.SAMN05934573    | family | order  |\n",
       "| type3b_family_3_0.33_0000_1502.SAMN05323905    | family | order  |\n",
       "| type3b_order_3_0.33_0000_115545.SAMN06140942   | order  | family |\n",
       "\n"
      ],
      "text/plain": [
       "   genome                                         contamination_level\n",
       "1  type3b_class_3_0.33_0000_1122947.SAMN01057329  class              \n",
       "2  type3b_class_3_0.33_0000_1123249.SAMN02440643  class              \n",
       "3  type3b_class_3_0.33_0000_1149862.SAMN01057330  class              \n",
       "4  type3b_class_3_0.33_0000_1160201.SAMN00811184  class              \n",
       "5  type3b_class_3_0.33_0000_1287313.SAMN02359689  class              \n",
       "6  type3b_class_3_0.33_0000_1298595.SAMD00000528  class              \n",
       "7  type3b_class_3_0.33_0000_1505.SAMEA1572114     class              \n",
       "8  type3b_family_3_0.33_0000_1278077.SAMN02469985 family             \n",
       "9  type3b_family_3_0.33_0000_1428.SAMN04278951    family             \n",
       "10 type3b_family_3_0.33_0000_1433.SAMN05934573    family             \n",
       "11 type3b_family_3_0.33_0000_1502.SAMN05323905    family             \n",
       "12 type3b_order_3_0.33_0000_115545.SAMN06140942   order              \n",
       "   gtdb_contamination_level\n",
       "1  phylum                  \n",
       "2  phylum                  \n",
       "3  phylum                  \n",
       "4  phylum                  \n",
       "5  phylum                  \n",
       "6  phylum                  \n",
       "7  phylum                  \n",
       "8  genus                   \n",
       "9  order                   \n",
       "10 order                   \n",
       "11 order                   \n",
       "12 family                  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# do we get the 11 genomes back where we found a mismatch before?\n",
    "\n",
    "reset_contam_level2 %>%\n",
    "  filter(! contamination_level %in% c(\"genus\", \"specI\")) %>% # we don't expect to do good here anyway, just filter for now\n",
    "  filter(contamination_level != gtdb_contamination_level)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "72ab8d52",
   "metadata": {},
   "outputs": [],
   "source": [
    "write_tsv(reset_contam_level2, \"sandbox/gunc_synthetic_genomes3b/type3b.genomes/genome_contamination_level.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fbf865e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
